netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -258,6 +261,40 @@ _ebpf_sock_addr_get_current_logon_id(_In_ const bpf_sock_addr_t* ctx)
     return logon_id;
 }
 
+static int
+_ebpf_sock_addr_set_redirect_context(_In_ const bpf_sock_addr_t* ctx, _In_ void* data, _In_ uint32_t data_size)
+{
+    int return_value = 0;
+    net_ebpf_sock_addr_t* sock_addr_ctx = CONTAINING_RECORD(ctx, net_ebpf_sock_addr_t, base);
+    void* redirect_context = NULL;
+
+    // This function is only supported at the connect redirect layer.
+    if (sock_addr_ctx->hook_id != EBPF_HOOK_ALE_CONNECT_REDIRECT_V4 &&
+        sock_addr_ctx->hook_id != EBPF_HOOK_ALE_CONNECT_REDIRECT_V6) {
+        return_value = -1;
+        goto Exit;
+    }
+
+    // Allocate buffer to store redirect context.
+    redirect_context = ExAllocatePoolUninitialized(NonPagedPoolNx, data_size, NET_EBPF_EXTENSION_POOL_TAG);
+    if (redirect_context == NULL) {
+        return_value = -1;
+        goto Exit;
+    }
+    memcpy(redirect_context, data, data_size);
+
+    // Set redirect context.
+    sock_addr_ctx->redirect_context = redirect_context;
```

Comment:
what if the ebpf program calls this helper function more than once? We will leak memory in that case I think.

Reply:
Do you have thoughts on the behavior here? Should we return an error if already set or should we replace the context with the new one?

Reply:
Updated this change to free any existing context if set.